File: js/InfluenceMap.js
//Translated from influence_map.rb using babyruby2js
'use strict';

var main = require('./main');
var Grid = require('./Grid');

/** @class public read-only attribute: map
 */
function InfluenceMap(goban) {
    var self = this;
    this.goban = goban;
    this.gsize = goban.gsize;
    this.map = Array.new(this.gsize + 1, function () {
        return Array.new(self.gsize + 1, function () {
            return [0, 0];
        });
    });
}
module.exports = InfluenceMap;

InfluenceMap.prototype.clear = function () {
    for (var j = 1; j <= this.gsize; j++) {
        for (var i = 1; i <= this.gsize; i++) {
            for (var c = 0; c < 2; c++) {
                this.map[j][i][c] = 0;
            }
        }
    }
};

InfluenceMap.prototype.buildMap = function () {
    this.clear();
    var influence = [4, 2, 1];
    // First we get stones' direct influence
    for (var j = 1; j <= this.gsize; j++) {
        for (var i = 1; i <= this.gsize; i++) {
            var stone = this.goban.stoneAt(i, j);
            var color = stone.color;
            if (color !== main.EMPTY) {
                this.map[j][i][color] += influence[0];
                // Then we propagate it decreasingly with distance
                for (var n1, n1_array = stone.neighbors, n1_ndx = 0; n1=n1_array[n1_ndx], n1_ndx < n1_array.length; n1_ndx++) {
                    if (n1.color !== main.EMPTY) {
                        continue;
                    }
                    this.map[n1.j][n1.i][color] += influence[1];
                    // Second level
                    for (var n2, n2_array = n1.neighbors, n2_ndx = 0; n2=n2_array[n2_ndx], n2_ndx < n2_array.length; n2_ndx++) {
                        if (n2.color !== main.EMPTY) {
                            continue;
                        }
                        if (n2 === stone) {
                            continue;
                        }
                        this.map[n2.j][n2.i][color] += influence[2]; // 3rd level // n2.neighbors.each do |n3| //   next if n3 == n1 //   @map[n3.j][n3.i][color] += influence[3] // end
                    }
                }
            }
        }
    }
    if (main.debug) {
        return this.debugDump();
    }
};

InfluenceMap.prototype.debugDump = function () {
    var c;
    function inf2str(inf) { return '%2d'.format(inf[c]); }

    for (c = 0; c < 2; c++) {
        console.log('Influence map for ' + Grid.COLOR_NAMES[c] + ':');
        for (var j = this.gsize; j >= 1; j--) {
            console.log('' + '%2d'.format(j) +
                this.map[j].slice(1, this.gsize + 1).map(inf2str).join('|'));
        }
        var cols = '  ';
        for (var i = 1; i <= this.gsize; i++) { cols += ' ' + Grid.xLabel(i) + ' '; }
        console.log(cols);
    }
};

File: rb/influence_map.rb

class InfluenceMap

  attr_reader :map

  def initialize(goban)
    @goban = goban
    @gsize = goban.gsize
    @map = Array.new(@gsize+1) {Array.new(@gsize+1) {[0,0]}}
  end
  
  def clear
    1.upto(@gsize) do |j|
      1.upto(@gsize) do |i|
        2.times { |c| @map[j][i][c] = 0 }
      end
    end
  end
  
  def build_map
    clear
    influence = [4,2,1]
    # First we get stones' direct influence
    1.upto(@gsize) do |j|
      1.upto(@gsize) do |i|
        stone = @goban.stone_at?(i,j)
        color = stone.color
        if color != EMPTY
          @map[j][i][color] += influence[0]
          # Then we propagate it decreasingly with distance
          stone.neighbors.each do |n1|
            next if n1.color != EMPTY
            @map[n1.j][n1.i][color] += influence[1]
            # Second level
            n1.neighbors.each do |n2|
              next if n2.color != EMPTY
              next if n2 == stone
              @map[n2.j][n2.i][color] += influence[2]
              # 3rd level
              # n2.neighbors.each do |n3|
              #   next if n3 == n1
              #   @map[n3.j][n3.i][color] += influence[3]
              # end
            end
          end
        end
      end
    end
    debug_dump if $debug
  end
  
  def debug_dump
    2.times do |c|
      puts "Influence map for #{Grid::COLOR_NAMES[c]}:"
      @gsize.downto(1) do |j|
        print "#{'%2d' % j}"
        1.upto(@gsize) do |i|
          print "#{'%2d' % @map[j][i][c]}|"
        end
        print "\n"
      end
      print "  "
      1.upto(@gsize) { |i| print " #{Grid.x_label(i)} " }
      print "\n"
    end
  end

end

